File: leetcode/0355-design-twitter/0355-design-twitter.js
var Twitter = function() {
    this.following = new Map();
    this.tweets = new Map();
    this.time = 0;
};

Twitter.prototype.postTweet = function(userId, tweetId) {
    if (!this.tweets.has(userId)) {
        this.tweets.set(userId, []);
    }

    this.tweets.get(userId).push([
        this.time++,
        tweetId
    ]);
};

Twitter.prototype.getNewsFeed = function(userId) {
    const result = [];
    const maxHeap = [];

    const push = (tweet) => {
        maxHeap.push(tweet);

        let i = maxHeap.length - 1;

        while (i > 0) {
            const parent = Math.floor((i - 1) / 2);

            if (maxHeap[parent][0] >= maxHeap[i][0]) {
                break;
            }

            [maxHeap[parent], maxHeap[i]] =
                [maxHeap[i], maxHeap[parent]];

            i = parent;
        }
    };

    const pop = () => {
        const top = maxHeap[0];
        const last = maxHeap.pop();

        if (maxHeap.length > 0) {
            maxHeap[0] = last;

            let i = 0;

            while (true) {
                let largest = i;
                const left = i * 2 + 1;
                const right = i * 2 + 2;

                if (
                    left < maxHeap.length &&
                    maxHeap[left][0] > maxHeap[largest][0]
                ) {
                    largest = left;
                }

                if (
                    right < maxHeap.length &&
                    maxHeap[right][0] > maxHeap[largest][0]
                ) {
                    largest = right;
                }

                if (largest === i) {
                    break;
                }

                [maxHeap[i], maxHeap[largest]] =
                    [maxHeap[largest], maxHeap[i]];

                i = largest;
            }
        }

        return top;
    };

    // 자신의 트윗
    if (this.tweets.has(userId)) {
        const list = this.tweets.get(userId);

        for (
            let i = Math.max(0, list.length - 10);
            i < list.length;
            i++
        ) {
            push(list[i]);
        }
    }

    // 팔로우한 사용자의 트윗
    const followees =
        this.following.get(userId) || new Set();

    for (const followee of followees) {
        if (!this.tweets.has(followee)) {
            continue;
        }

        const list = this.tweets.get(followee);

        for (
            let i = Math.max(0, list.length - 10);
            i < list.length;
            i++
        ) {
            push(list[i]);
        }
    }

    // 최신 트윗부터 최대 10개
    while (maxHeap.length > 0 && result.length < 10) {
        result.push(pop()[1]);
    }

    return result;
};

Twitter.prototype.follow = function(followerId, followeeId) {
    if (!this.following.has(followerId)) {
        this.following.set(followerId, new Set());
    }

    this.following.get(followerId).add(followeeId);
};

Twitter.prototype.unfollow = function(followerId, followeeId) {
    if (this.following.has(followerId)) {
        this.following.get(followerId).delete(followeeId);
    }
};

File: leetcode/0355-design-twitter.java
import java.util.*;

class Twitter {
    private final Map<Integer, Set<Integer>> following;
    private final Map<Integer, List<int[]>> tweets;
    private int time;

    public Twitter() {
        following = new HashMap<>();
        tweets = new HashMap<>();
        time = 0;
    }

    public void postTweet(int userId, int tweetId) {
        tweets.putIfAbsent(userId, new ArrayList<>());
        tweets.get(userId).add(new int[]{time++, tweetId});
    }

    public List<Integer> getNewsFeed(int userId) {
        List<Integer> result = new ArrayList<>();

        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a, b) -> Integer.compare(b[0], a[0])
        );

        // 자기 자신의 트윗
        if (tweets.containsKey(userId)) {
            List<int[]> list = tweets.get(userId);
            for (int i = Math.max(0, list.size() - 10); i < list.size(); i++) {
                pq.offer(list.get(i));
            }
        }

        // 팔로우 중인 사용자들의 트윗
        Set<Integer> followees =
            following.getOrDefault(userId, Collections.emptySet());

        for (int followee : followees) {
            if (!tweets.containsKey(followee)) {
                continue;
            }

            List<int[]> list = tweets.get(followee);

            for (int i = Math.max(0, list.size() - 10); i < list.size(); i++) {
                pq.offer(list.get(i));
            }
        }

        while (!pq.isEmpty() && result.size() < 10) {
            result.add(pq.poll()[1]);
        }

        return result;
    }

    public void follow(int followerId, int followeeId) {
        following.putIfAbsent(followerId, new HashSet<>());
        following.get(followerId).add(followeeId);
    }

    public void unfollow(int followerId, int followeeId) {
        if (following.containsKey(followerId)) {
            following.get(followerId).remove(followeeId);
        }
    }
}